### PR TITLE
[WIP][HIVEMALL-99] Cross-compilation of XGBoost using Docker

### PR DESCRIPTION
**Pull request by amaya382**
## What changes were proposed in this pull request?

Cross compiling xgboost using dockcross

## What type of PR is it?

Improvement

## What is the Jira issue?

[[HIVEMALL-99]](https://issues.apache.org/jira/browse/HIVEMALL-99)

## How was this patch tested?

manually by qemu

## How to use this feature?

```sh
JAVA_HOME=/etc/alternatives/java_sdk make xgboost-native-local # on local
make xgboost-native-linux-x64 # for specific arch (`make xgboost-native-${arch}`)
make xgboost-native # all arches
```
out
```sh
${HIVEMALL_HOME}/xgboost/src/main/resources/lib/${arch}/libxgboost4j.so
```

## Supported arches

* linux/x64
* linux/x86 (Note: [x86 is NOT officially supported by xgboost](https://github.com/dmlc/xgboost/issues/1807#issuecomment-262613978))
* linux/arm64
* linux/ppc64le (future work?)
* ~~windows~~ (future work)
* ~~macOS~~ (not supported by dockcross, but expected to work well)

## Tasks

- [x] Build on local
- [x] Build on docker (dockcross)
- [x] Rebase master (for #95)
- Test by QEMU
  - [x] linux/x64
  - [ ] linux/x86
  - [x] linux/arm64
  - [ ] linux/ppc64le

### CODE REVIEW

**Comment by amaya382** [comment]
@myui @maropu 
* How to distribute each product?
* How to test them on CI?

Other topic
Would you like to move build scripts into Makefile?

**Comment by myui** [comment]
@amaya382 Fist of all, thank you for the contribution!

Could you share the result of `ldd libxgboost4j.so` ?

We would like to create a `libxgboost4j.so` that runs on many linux system by using `-fPIC` option and static linking. 
https://github.com/xerial/snappy-java/blob/master/BUILD.md#building-linux-x86_64-binary
https://github.com/xerial/snappy-java/blob/master/Makefile#L169

**Comment by amaya382** [comment]
@myui 
Ahh, I've got it! will fix it (currently not static-linked).

**Comment by myui** [comment]
@amaya382  BTW, what's the result of `ldd libxgboost4j.so` ?

**Comment by amaya382** [comment]
@myui 
now
```
linux-vdso.so.1 (0x00007ffdfc7cc000)
librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fce0f7f6000)
libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fce0f4f5000)
libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fce0f2d7000)
libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fce0ef2c000)
/lib64/ld-linux-x86-64.so.2 (0x0000555866c1f000)
```

`libgcc` and `libstdc++` are statically linked with `fPIC`. but still `libc` is dynamic.

**Comment by myui** [comment]
@amaya382 having dependencies to `libc` and `libm` is acceptable.
https://github.com/xerial/snappy-java/issues/17

**Comment by amaya382** [comment]
@myui okay. what should I set default arch in [here](https://github.com/apache/incubator-hivemall/pull/80/files#diff-600376dffeb79835ede4a0b285078036) or provide multiple binaries?

**Comment by myui** [comment]
@amaya382 considering to build multiple shared libraries in resource directory as seen in 
https://github.com/xerial/snappy-java/tree/master/src/main/resources/org/xerial/snappy/native

**Comment by myui** [comment]
I would like to update xgboost version to latest using their official repository.

@maropu I think it's better to use jitpack for jvm-packages of xgboost if maven repository is required.

https://github.com/jitpack/maven-simple 
https://github.com/myui/digdag-plugin-example/

**Comment by maropu** [comment]
We need to have these binaries for older linux platforms? I think xgboost actually does not work on those older ones. Also, we need to compile xgboost for some platforms (e.g., arm and some bigendian ones). That is because Spark does not work well on those platforms: See: https://issues.apache.org/jira/browse/SPARK-13736. I'm not sure though, Hive works on those?

File: bin/build_xgboost.sh [hunk 1]
```diff
@@ -1,87 +0,0 @@
...
-# under the License.
-#
-
-# xgboost requires g++-4.6 or higher (https://github.com/dmlc/xgboost/blob/master/doc/build.md),
-# so we need to first check if the requirement is satisfied.
-COMPILER_REQUIRED_VERSION="4.6"
-COMPILER_VERSION=`g++ --version 2> /dev/null`
-
-# Check if GNU g++ installed
-if [ $? = 127 ]; then
```
Comment (maropu) [inline bin/build_xgboost.sh]: @amaya382 Does the new script works for both gcc and clang?

File: xgboost/src/main/java/hivemall/xgboost/NativeLibLoader.java [hunk 2]
```diff
@@ -54,15 +55,47 @@ private static boolean hasResource(String path) {
...
+            arch = "arm64";
+        } else if(arch.indexOf("armv6") != -1) {
+            arch = "armv6";
+        } else if(arch.indexOf("armv7") != -1) {
+            arch = "armv7";
+        } else if(arch.indexOf("ppc") != -1) {
+            arch = "ppc64le";
+        }
+
+        return os + "-" + arch;
```
Comment (maropu) [inline xgboost/src/main/java/hivemall/xgboost/NativeLibLoader.java]: I think you could refer [the code](https://github.com/xerial/snappy-java/blob/master/src/main/java/org/xerial/snappy/OSInfo.java) in `snappy-java` to handle almost all the case for detecting arch.

**Comment by maropu** [comment]
@amaya382 I think it's hard to test these binaries on the current Travis. So, we must manually tests them before these binaries are included.

File: bin/build_xgboost.sh (hunk 1 above)
Comment (amaya382) [inline bin/build_xgboost.sh]: @maropu No, building xgboost depends on gcc. because dockcross includes gcc, skips to check compiler.

File: xgboost/src/main/java/hivemall/xgboost/NativeLibLoader.java (hunk 2 above)
Comment (amaya382) [inline xgboost/src/main/java/hivemall/xgboost/NativeLibLoader.java]: that's so, I'll mod it

**Comment by amaya382** [comment]
@maropu by qemu? or do you have any ideas?

File: bin/build_xgboost.sh (hunk 1 above)
Comment (maropu) [inline bin/build_xgboost.sh]: We'd be better off printing explicit error messages when `clang` used. Could you?

Comment (myui) [inline bin/build_xgboost.sh]: @maropu no need to support `clang` anymore.

**Comment by maropu** [comment]
Yea, I also think we need to use `qemu` to test them.

**Comment by myui** [comment]
Some ASF projects are using appveyor CI for windows testing. FYI

https://blogs.apache.org/infra/entry/appveyor_ci_now_available_for
https://github.com/apache/thrift/blob/master/appveyor.yml

File: bin/build_xgboost.sh (hunk 1 above)
Comment (amaya382) [inline bin/build_xgboost.sh]: this script uses `gcc` automatically on dockcross, never call `clang`

Comment (maropu) [inline bin/build_xgboost.sh]: Ah, ok. But I think we keep a script to build `xgboost` on native environments in terms of CPU optimization ( I think it'd be better to follow the same approach with `snappy-java`).

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#
+
+DOCKCROSS_SCRIPT := 'dockcross.bash'
+XGBOOST_REPO := 'https://github.com/maropu/xgboost.git'
+XGBOOST_BRANCH := 'xgboost_v0.60_with_portable_binaries'
+HIVEMALL_HOME := "$(shell pwd)"
+HIVEMALL_OUT := "${HIVEMALL_HOME}/target"
+XGBOOST_OUT := "${HIVEMALL_OUT}/xgboost"
+HIVEMALL_LIB_DIR := "${HIVEMALL_HOME}/xgboost/src/main/resources/lib"
+CANDIDATES := 'linux-arm64' 'linux-armv6' 'linux-armv7' 'linux-ppc64le' 'linux-x64' 'linux-x86' 'windows-x64' 'windows-x86'
```
Comment (myui) [inline Makefile]: supported platform of hivemall depends on Hadoop.

target os is linux, windows, mac osx.
main target platform is x86, x64.

mac osx support is only used for local testing though.
https://github.com/xerial/snappy-java/blob/master/Makefile#L167
https://github.com/multiarch/crossbuild

**Comment by amaya382** [comment]
@maropu This isn't directly related with this PR, but I faced the following errors. Do you have any ideas? It seems that labels aren't set properly.

Env: Docker, xgboost native built on local

Commands in hive
```sql
add jar /opt/hivemall/target/hivemall-core-0.4.2-rc.2-with-dependencies.jar;
source /opt/hivemall/resources/ddl/define-all.hive;
add jar /opt/hivemall/target/hivemall-xgboost-0.60-0.4.2-rc.2-with-dependencies.jar;
source /opt/hivemall/resources/ddl/define-additional.hive;

set hivevar:f0_min=4.3;
set hivevar:f0_max=7.9;
set hivevar:f1_min=2.0;
set hivevar:f1_max=4.4;
set hivevar:f2_min=1.0;
set hivevar:f2_max=6.9;
set hivevar:f3_min=0.1;
set hivevar:f3_max=2.5;

use iris;
create or replace view iris_scaled
as
select
  rowid, 
  label,
  add_bias(array(
     concat("1:", rescale(features[0],${hivevar:f0_min},${hivevar:f0_max})), 
     concat("2:", rescale(features[1],${hivevar:f1_min},${hivevar:f1_max})), 
     concat("3:", rescale(features[2],${hivevar:f2_min},${hivevar:f2_max})), 
     concat("4:", rescale(features[3],${hivevar:f3_min},${hivevar:f3_max}))
  )) as features
from 
  iris_raw;

-- select * from iris_scaled limit 3;
-- 1       Iris-setosa     ["1:0.22222215","2:0.625","3:0.0677966","4:0.041666664","0:1.0"]
-- 2       Iris-setosa     ["1:0.16666664","2:0.41666666","3:0.0677966","4:0.041666664","0:1.0"]
-- 3       Iris-setosa     ["1:0.11111101","2:0.5","3:0.05084745","4:0.041666664","0:1.0"]

select train_xgboost_classifier(features, case when label = 'Iris-setosa' then 1.0 else 0.0 end) from iris_scaled; -- got exception
```

```
Failed with exception java.io.IOException:java.lang.ClassCastException: org.apache.hadoop.io.Text cannot be cast to java.lang.String
[20:51:33] dmlc-core/include/dmlc/logging.h:235: [20:51:33] src/objective/regression_obj.cc:89: Check failed: (info.labels.size()) != (0) label set cannot be empty
org.apache.hadoop.hive.ql.metadata.HiveException: [20:51:33] src/objective/regression_obj.cc:89: Check failed: (info.labels.size()) != (0) label set cannot be empty
Check failed: (info.labels.size()) != (0) label set cannot be empty
        at hivemall.xgboost.XGBoostUDTF.close(XGBoostUDTF.java:313)
        at org.apache.hadoop.hive.ql.exec.UDTFOperator.closeOp(UDTFOperator.java:152)
        at org.apache.hadoop.hive.ql.exec.Operator.close(Operator.java:683)
        at org.apache.hadoop.hive.ql.exec.Operator.close(Operator.java:697)
        at org.apache.hadoop.hive.ql.exec.Operator.close(Operator.java:697)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.closeOperator(FetchOperator.java:552)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.clearFetchContext(FetchOperator.java:535)
        at org.apache.hadoop.hive.ql.exec.FetchTask.clearFetch(FetchTask.java:191)
        at org.apache.hadoop.hive.ql.Driver.releaseFetchTask(Driver.java:2233)
        at org.apache.hadoop.hive.ql.Driver.close(Driver.java:2278)
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:269)
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:183)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:399)
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:776)
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:714)
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:641)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
org.apache.hadoop.hive.ql.metadata.HiveException: [20:51:33] src/objective/regression_obj.cc:89: Check failed: (info.labels.size()) != (0) label set cannot be empty
        at hivemall.xgboost.XGBoostUDTF.close(XGBoostUDTF.java:313)
        at org.apache.hadoop.hive.ql.exec.UDTFOperator.closeOp(UDTFOperator.java:152)
        at org.apache.hadoop.hive.ql.exec.Operator.close(Operator.java:683)
        at org.apache.hadoop.hive.ql.exec.Operator.close(Operator.java:697)
        at org.apache.hadoop.hive.ql.exec.Operator.close(Operator.java:697)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.closeOperator(FetchOperator.java:552)
        at org.apache.hadoop.hive.ql.exec.FetchOperator.clearFetchContext(FetchOperator.java:535)
        at org.apache.hadoop.hive.ql.exec.FetchTask.clearFetch(FetchTask.java:191)
        at org.apache.hadoop.hive.ql.Driver.releaseFetchTask(Driver.java:2233)
        at org.apache.hadoop.hive.ql.Driver.close(Driver.java:2278)
        at org.apache.hadoop.hive.cli.CliDriver.processLocalCmd(CliDriver.java:269)
        at org.apache.hadoop.hive.cli.CliDriver.processCmd(CliDriver.java:183)
        at org.apache.hadoop.hive.cli.CliDriver.processLine(CliDriver.java:399)
        at org.apache.hadoop.hive.cli.CliDriver.executeDriver(CliDriver.java:776)
        at org.apache.hadoop.hive.cli.CliDriver.run(CliDriver.java:714)
        at org.apache.hadoop.hive.cli.CliDriver.main(CliDriver.java:641)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.apache.hadoop.util.RunJar.run(RunJar.java:221)
        at org.apache.hadoop.util.RunJar.main(RunJar.java:136)
Time taken: 3.375 seconds
```

**Comment by maropu** [comment]
@amaya382 Aha, I've not seen that exception. Actually, I didn't check behaviours in Hive. Could you look into this issue?

**Comment by myui** [comment]
@maropu It seems it's not hive specific issue but not label set is the training dataset.

**Comment by myui** [comment]
`$ JAVA_HOME=`/usr/libexec/java_home -v 1.8` make xgboost-native-local`
`javac: ディレクトリがありません: xgboost/target/classes`

Need to run `mkdir -p xgboost/target/classes` first.

**Comment by maropu** [comment]
yea, I think so. I just mean I can't reproduce in my laptop and I can't look into this issue...

**Comment by myui** [comment]
@maropu you can use Docker.

```
docker pull hivemall/latest:20170517
docker run -p 8088:8088 -p 50070:50070 -p 19888:19888 -it hivemall/latest:20170517
```

**Comment by amaya382** [comment]
@maropu I'll investigate it too, but want you to check it because i don't know xgboost in detail.
In addition to myui's comment, increased my environment info into previous comment.

**Comment by amaya382** [comment]
@myui thanks for reporting bugs, fixed first bug. it seems second bug is problem related to first bug or particular to macOS. plz fetch the latest and retry. if former, will be resolved.

**Comment by maropu** [comment]
@amaya382 Could you file a jira for that?

**Comment by amaya382** [comment]
@maropu Created https://issues.apache.org/jira/projects/HIVEMALL/issues/HIVEMALL-119

**Comment by myui** [comment]
`git fetch; git rebase master` is required to resolve this conflict.

**Comment by myui** [comment]
@amaya382 Sorry to be late. https://github.com/apache/incubator-hivemall/pull/95 is now merged. So, please rebase and test again.

**Comment by myui** [comment]
@amaya382 Thanks! Coverage/covealls decrease is acceptable for this PR. Is it still WIP?

**Comment by amaya382** [comment]
@myui It has been tested only on local, plz wait a little

**Comment by myui** [comment]
sure.

**Comment by amaya382** [comment]
@myui Sorry for the work late, is this PR no longer necessary or not?

**Comment by myui** [comment]
@amaya382 Instead of using dockcross, I built native xgboost binaries for most Linux platform and Mac OSX in https://github.com/myui/build-xgboost-jvm/ 

To support more platforms, this PR is useful but we can close this PR for now. Thank you for your work but do you mind closing/archiving this PR for the time being?

**Comment by amaya382** [comment]
@myui I see
No. If need be, I'll open it as a new PR
